### ISSUE TITLE
[$250] Web - Compose Box - Emoji only is being accepted as an anchor text

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#54484]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.78-1
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught on HybridApp, is this reproducible on New Expensify Standalone?:** N/A
**If this was caught during regression testing, add the test name, ID and link from TestRail:** N/A
**Issue reported by:** Applause - Internal Team

## Action Performed:
1. Go to https://staging.new.expensify.com/
2. Open a chat and put an emoji inside of square brackets "[]" and put a link inside of parenthesis "()" and post the comment

## Expected Result:
Emoji only should not be accepted as an anchor text

## Actual Result:
Emoji only is being accepted as an anchor text for hyperlinks

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Standalone
- [ ] Android: HybridApp
- [ ] Android: mWeb Chrome
- [ ] iOS: Standalone
- [ ] iOS: HybridApp
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

<details>
  <summary>Add any screenshot/video evidence</summary>

  
</details> 

https://github.com/user-attachments/assets/d13b8782-e539-4bc5-8e3d-bfd71f7964bf

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021871677968650628808</li>
        <li>Upwork Job ID: 1871677968650628808</li>
        <li>Last Price Increase: 2024-12-24</li>
    </ul>
</details>

**Comment by sakluger** [comment]
We fixed this a long time ago here: https://github.com/Expensify/App/issues/18386

@dukenv0307 and @sobitneupane worked on that issue. Do either of you have any ideas where this regression came from?

**Comment by varun531** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

The issue appears to be with hyperlink formatting in the chat functionality. When a user attempts to create a link with only an emoji as the anchor text ( using markdown format \[emoji\](url)), the app is accepting this as valid when it should not. The core problem seems to be insufficient validation of anchor text in the links.

### What is the root cause of that problem?

The `updateComment` in `ComposerWithSuggestions` processes emojis but doesn't validate the structural integrity of markdown elements like links.

### What changes do you think we should make in order to solve the problem?

Add a utility function to check for valid anchor text, maybe in `EmojiUtils`:

```typescript
const validateEmojiLinks = (text: string): boolean => {
  const linkRegex = /\[([^\]]+)\]\(([^)]+)\)/g;
  let match;
  
  while ((match = linkRegex.exec(text)) !== null) {
    const [, linkText] = match;
    // Check if the anchor text contains only emojis
    const nonEmojiContent = linkText.replace(CONST.REGEX.EMOJI_REGEX, '').trim();
    if (nonEmojiContent.length === 0) {
      return false; // Invalid - contains only emojis
    }
  }
  
  return true; // Valid - no emoji-only links found
};
```

Add this to the start of the `updateComment` method in `ComposerWithSuggestions`. It will convert emoji only links to: {emoji} {url}:

```typescript
const updateComment = useCallback(
    (commentValue: string, shouldDebounceSaveComment?: boolean) => {
        raiseIsScrollLikelyLayoutTriggered();
        
        let validatedComment = commentValue;
        if (!validateEmojiLinks(commentValue)) {
            // If emoji-only links found, convert [emoji](url) to "emoji url"
            const linkRegex = /\[([^\]]+)\]\(([^)]+)\)/g;
            validatedComment = commentValue.replace(linkRegex, (fullMatch, anchorText, url) => {
                return `${anchorText} ${url}`;
            });
        }
        
        // Continue with existing emoji processing
        const {startIndex, endIndex, diff} = findNewlyAddedChars(lastTextRef.current, validatedComment);
```

This is obviously a rough attempt, we'd probably add another function to `EmojiUtils` to do the replacement, or combine validation and replacement into one method.

### What specific scenarios should we cover in automated tests to prevent reintroducing this issue in the future?


You could obviously add unit tests for verifying the validation function.

### What alternative solutions did you explore? (Optional)

In `onChangeText` in `ComposerWithSuggestions`, you could not call `updateComment` if this problematic link is detected.

**Comment by varun531** [comment]
Contributor details
Your Expensify account email: varun531@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~016c3e14bfc62eb770

**Comment by sakluger** [comment]
Thanks @dukenv0307! That's super helpful context. It seems like we started allowing emojis to be hyperlinked because they used to look bad (the underline intersected with the emoji), but they look fine now.

I don't see any issue with allowing hyperlinks when an emoji is part of a text thread, but I can see an argument for not allowing emoji-only hyperlinks. I started a discussion in Slack ([here](https://expensify.slack.com/archives/C06ML6X0W9L/p1735579454408309)) to see if we want to allow emoji-only hyperlinks or now.
